src/lib/optimizer/strategy/duplicate_elimination_rule.hpp

```diff
@@ -0,0 +1,35 @@
+#pragma once
+
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "abstract_rule.hpp"
+#include "logical_query_plan/abstract_lqp_node.hpp"
+#include "types.hpp"
+
+namespace opossum {
+
+class AbstractLQPNode;
+
+/**
+ * This optimizer rule eliminates duplicates of the LQP by using sub-plan memoization:
+ * This process uses depth-first traversal and starts integrating sub-plans into the
+ * memoization structure at the leaves of the overall LQP and works upwards to the root.
+ */
+
+class DuplicateEliminationRule : public AbstractRule {
+ public:
+  std::string name() const override;
+  void apply_to(const std::shared_ptr<AbstractLQPNode>& node) const override;
+
+ protected:
+  void _find_sub_plan_duplicates_traversal(const std::shared_ptr<AbstractLQPNode>& node) const;
+  void _adapt_expressions_traversal(const std::shared_ptr<AbstractLQPNode>& node) const;
+  mutable std::vector<std::shared_ptr<AbstractLQPNode>> _remaining_stored_table_nodes;
```

Comment:
Optimizer might run on multiple plans concurrently, don't store something here, put it into the functions themselves